ddtrace/internal/runtime/runtime_metrics.py

```diff
@@ -50,32 +53,81 @@ class RuntimeMetrics(RuntimeCollectorsIterable):
     ]
 
 
-@attr.s
+@attr.s(eq=False)
 class RuntimeWorker(periodic.PeriodicService):
     """Worker thread for collecting and writing runtime metrics to a DogStatsd
     client.
     """
 
-    dogstatsd_url = attr.ib(type=str)
     _interval = attr.ib(type=float, factory=lambda: float(get_env("runtime_metrics", "interval", default=10)))
+    tracer = attr.ib(type=Optional[ddtrace.Tracer], default=None)
```

Comment:
```suggestion
    tracer = attr.ib(type=ddtrace.Tracer, factory=lambda: ddtrace.tracer)
```

Reply:
If I do it this way, then every time `None` is passed in for `tracer`, `self.tracer` would be set to `None` rather than to `ddtrace.tracer` :(

Reply:
Right, though that sounds exactly like it'd be designed for. Then it's a misuse of the API and a violation of typing, but why are you trying to prevent this?

Reply:
This pattern, i.e. the use of `None` to signal a default behaviour, is used in other places AFAIK, so this is just in line with that. The type violation is just a consequence of using `attr`, since it's tying constructor arguments with instance attributes. That's why I'm not a big fan of using `attr` with objects other than actual data classes :(.

Reply:
Understood.

I think the use of `None` as a placeholder for default is a common, well-working, abuse of Python kwargs where it should rather be:

```python
_undefined = object()

def x(value=_undefined):
  if value is _undefined:
     value = some_default
```

`None` isn't really used as a default placeholder but has a convenient way of saying "if you didn't pass the kwarg then we'll detect it by checking for `None` being the value". Since you never expect users to pass `None` as a valid value because your type is not `Optional`, that works most of the time — even it's a little abusive.

To circle back to this, that's why I don't think it's a real problem. Not passing a kwarg is valid and gives you the correct result. Passing `None` explicitly is invalid (break typing), like passing `1234` or `"foobar"`.